boot/bootstate20.go

```diff
@@ -21,329 +21,511 @@ package boot
-		}
+	if bs20.modeenv.Base == "" {
+		return nil, nil, "", fmt.Errorf("cannot get snap revision: modeenv base boot variable is empty")
+	}
 
-		// clean
-		kernelStatus = ""
-		rebootRequired = false
-	} else {
-		// need to enable the symlink
-		u20.snapToTry = nextKernel
+	bootSn, err = snap.ParsePlaceInfoFromSnapFileName(bs20.modeenv.Base)
+	if err != nil {
+		return nil, nil, "", fmt.Errorf("cannot get snap revision: modeenv base boot variable is invalid: %v", err)
 	}
 
-	u20.toCommit["kernel_status"] = kernelStatus
+	if bs20.modeenv.BaseStatus == TryingStatus && bs20.modeenv.TryBase != "" {
+		tryBootSn, err = snap.ParsePlaceInfoFromSnapFileName(bs20.modeenv.TryBase)
+		if err != nil {
+			return nil, nil, "", fmt.Errorf("cannot get snap revision: modeenv try base boot variable is invalid: %v", err)
+		}
+	}
 
-	return rebootRequired, u20, nil
+	return bootSn, tryBootSn, bs20.modeenv.BaseStatus, nil
 }
 
-type bootStateUpdate20 struct {
-	typ       snap.Type
-	ebl       bootloader.ExtractedRunKernelImageBootloader
-	env       map[string]string
-	toCommit  map[string]string
-	snapToTry snap.PlaceInfo
-	snapTried snap.PlaceInfo
+func (bs20 *bootState20Base) setNext(next snap.PlaceInfo) (rebootRequired bool, u bootStateUpdate, err error) {
+	nextStatus, err := genericSetNext(bs20, next)
+	if err != nil {
+		return false, nil, err
+	}
+	// if we are setting a snap as a try snap, then we need to reboot
+	rebootRequired = false
+	if nextStatus == TryStatus {
+		bs20.tryBaseSnap = next
+		rebootRequired = true
+	}
+	bs20.commitBaseStatus = nextStatus
+
+	return rebootRequired, bs20, nil
 }
 
-func newBootStateUpdate20(u bootStateUpdate, typ snap.Type) (*bootStateUpdate20, error) {
-	// TODO:UC20: re-factor to handle base snap update as well as kernel snap
-	// TODO:UC20: handle base_status from modeenv as well
-	if u != nil {
-		u20, ok := u.(*bootStateUpdate20)
-		if !ok {
-			return nil, fmt.Errorf("internal error: threading unexpected boot state update on UC20: %T", u)
-		}
-		return u20, nil
-	}
-	bl, err := bootloader.Find("", nil)
+func (bs20 *bootState20Base) markSuccessful(update bootStateUpdate) (bootStateUpdate, error) {
+	// call the generic method with to do most of the legwork
+	u, sn, shouldClean, err := genericMarkSuccessful(bs20, update)
 	if err != nil {
 		return nil, err
 	}
-	ebl, ok := bl.(bootloader.ExtractedRunKernelImageBootloader)
-	if !ok {
-		return nil, fmt.Errorf("cannot use %s bootloader: does not support extracted run kernel images", bl.Name())
+
+	if shouldClean {
+		logger.Noticef("base_status is %q, but there is no try base snap, resetting to %q", bs20.modeenv.BaseStatus, DefaultStatus)
+		bs20.commitBaseStatus = DefaultStatus
+	}
+
+	// u should always be non-nil if err is nil
+	u.triedBaseSnap = sn
+	return u, nil
+}
+
+// commit for bootState20Base is meant only to be used with setNext(), for
+// markSuccessful(), use bootState20MarkSuccessful.
+func (bs20 *bootState20Base) commit() error {
```

Comment:
Had to switch to a local checkout for easier review. I think it'd be easier to follow if `commit()` was defined directly after `setNext()`.

Reply:
ok, will move